portal/flatpak-portal.c

```diff
@@ -1688,22 +1726,141 @@ get_app_display_name (const char *app_id)
   return g_strdup (app_id);
 }
 
+typedef enum {
+  UNINSTALLABLE,
+  SELF_UPDATE,
+  APP_EXTENSION,
+  RUNTIME_EXTENSION
+} UpdateKind;
+
+static UpdateKind
+classify_ref (PortalFlatpakUpdateMonitor *monitor,
+              const char                 *ref)
+{
+  UpdateMonitorData *m = update_monitor_get_data (monitor);
+  g_autoptr(GFile) installation_path = NULL;
+  g_autoptr(FlatpakInstallation) installation = NULL;
+  g_autoptr(FlatpakInstalledRef) installed_ref = NULL;
+  g_autoptr(FlatpakRef) r = NULL;
+  g_autoptr(FlatpakRemoteRef) rref = NULL;
+  GBytes *metadata = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+  const char *origin;
+  g_autofree char *extension_of = NULL;
+  g_autofree char *app_ref = NULL;
+
+  if (ref == NULL)
+    return SELF_UPDATE;
+
+  r = flatpak_ref_parse (ref, NULL);
+  if (flatpak_ref_get_kind (r) != FLATPAK_REF_KIND_RUNTIME)
+    return UNINSTALLABLE;
+
+  installation_path = update_monitor_get_installation_path (monitor);
+  installation = lookup_installation_for_path (installation_path, NULL);
+  if (installation == NULL)
+    return UNINSTALLABLE;
+
+  installed_ref = flatpak_installation_get_current_installed_app (installation, m->name, NULL, NULL);
+  if (installed_ref == NULL)
+    return UNINSTALLABLE;
+
+  /* TODO: handle different installation / origin between app and runtime */
+  origin = flatpak_installed_ref_get_origin (installed_ref);
+  rref = flatpak_installation_fetch_remote_ref_sync (installation,
+                                                     origin,
+                                                     flatpak_ref_get_kind (r),
+                                                     flatpak_ref_get_name (r),
+                                                     NULL, NULL,
+                                                     NULL, NULL);
+  if (!rref)
+    {
+      g_debug ("Did not find remote ref for %s in %s\n", ref, origin);
+      return UNINSTALLABLE;
+    }
+
+  metadata = flatpak_remote_ref_get_metadata (rref);
+  keyfile = g_key_file_new ();
+  if (!g_key_file_load_from_data (keyfile,
+                                  g_bytes_get_data (metadata, NULL),
+                                  g_bytes_get_size (metadata),
+                                  G_KEY_FILE_NONE, NULL))
+    return UNINSTALLABLE;
+
+  extension_of = g_key_file_get_string (keyfile,
+                                        FLATPAK_METADATA_GROUP_EXTENSION_OF,
+                                        FLATPAK_METADATA_KEY_REF,
+                                        NULL);
```

Comment:
This is not a great check. Many runtime extensions will be marked as extensions of the fd.o runtime, but the app might be using a derivate of it that still references the same extension point. I think we instead need to load the app/runtime metadata file and call flatpak_list_extensions() on that.

Reply:
ok